voxel_grid/prototype/mvp_v1.hpp

```diff
@@ -0,0 +1,226 @@
+    }
+
+    // Defined in pcl::PointCloud
+    using PointType = PointT;
+
+    std::vector<PointT> points;
+};
+
+template <typename PointT>
+class Filter
+{
+    using PointCloud = PointCloudT<PointXYZ>;
+
+public:
+    Filter() {}
+    virtual void applyFilter(PointCloud &pc) = 0;
+    void filter(PointCloud &output)
+    {
+        applyFilter(output);
+    }
+};
+
+// ================== Base of VoxelGrid ==================
+
+// Deduce type
+
+// I've tried deducing by: https://stackoverflow.com/questions/22632236/how-is-possible-to-deduce-function-argument-type-in-c
+// but it doesn't work in our case
+template <typename T>
+using get_point_type = typename T::PointCloud::PointType;
+
+template <typename T>
+using get_grid_type = typename T::Grid;
+
+template <typename GridStruct, typename Grid = get_grid_type<GridStruct>, typename PointT = get_point_type<GridStruct>>
+class GridFilter : public Filter<PointT>, public GridStruct
+{
+    // TODO:
+    // check if the required functions exist in Grid
+    // static_assert(...);
+
+    using PointCloud = PointCloudT<PointT>;
+
+public:
+    GridFilter() : GridStruct()
+    {
+    }
+
+    const PointCloud &getInputCloud() const
+    {
+        return input_;
+    }
+
+    void setInputCloud(const PointCloud &input)
+    {
+        input_ = input;
+    }
+
+    Eigen::Vector3f getLeafSize() const
+    {
+        return leaf_size_;
+    }
+
+    void getMinMax3D(const PointCloud &cloud, Vec3f &min_pt, Vec3f &max_pt) const
+    {
+        // TODO: set initialize values
+        for (const auto &pt : cloud.points)
+        {
+            min_pt = min_pt.cwiseMin(pt.p);
+            max_pt = max_pt.cwiseMax(pt.p);
+        }
+    }
+
+protected:
+    void applyFilter(PointCloud &output)
+    {
+        GridStruct::setUp(this);
+        GridStruct::addPointsToGrid(input_, grid_);
+
+        for (auto it = grid_.begin(); it != grid_.end(); ++it)
+            if (std::optional<PointT> res = GridStruct::filterGrid(it, grid_); res)
+                output.push_back(res.value());
+    }
+
+    PointCloud input_;
+    Grid grid_;
+    Eigen::Vector3f leaf_size_ = Eigen::Vector3f::Constant(0.05);
+    // ...
+};
+
+// ================== VoxelGrid.hpp ==================
+
+// structure of a single voxel
+struct Voxel
+{
+    Vec3f centroid = Vec3f::Zero();
+    size_t num_pt = 0;
+};
+
+// structure and operations of the grid
+template <typename PointT>
+class VoxelStructT
+{
+public:
+    // Require user to define them
+    // not sure if there are a nicer way to do it
+    using PointCloud = PointCloudT<PointT>;
+    using Grid = std::unordered_map<size_t, Voxel>;
+
+    void setMinimumPointsNumberPerVoxel(const size_t num)
+    {
+        min_points_per_voxel_ = num;
+    }
+
+protected:
+    void setUp(const GridFilter<VoxelStructT> *grid_filter)
+    {
+        Vec3f min_p, max_p;
+        const auto &input = grid_filter->getInputCloud();
+        grid_filter->getMinMax3D(input, min_p, max_p);
```

Comment:
Maybe use https://github.com/PointCloudLibrary/pcl/blob/853afb7b165e66c7ef730a452ab3d63fe81e979b/common/include/pcl/common/common.h#L189 ?

Reply:
Sure, I will use it when I implement the prototype inside pcl, the function there is just a mock to make the prototype running

Reply:
Ahh, true.